### ISSUE TITLE
Improve how analysis definitions are created

### ISSUE THREAD

**Issue by alonsogarciapablo** [CartoDB/cartodb#7608]
**TL;DR: Frontend is now taking care of creating analysis definitions when a layer or multiple layers are saved, but this could leave the app in an invalid state unless we implement some error handling or we make this operations atomic by moving some of the logic to the backend.**

Right now, the frontend takes care of creating analysis definitions so that layers that have a source point to an existing analysis definition. This is done in two places right now:

1. Whenever a layerDefinitionModel is synced (something that happens right after the layerDefinitionModel is saved), the frontend checks if the analysis definition needs to be created, in which case, it sends a POST request to the backend (to `api/v3/viz/VIS_ID/analyses`). Code is [here](https://github.com/CartoDB/cartodb/blob/94ee96f3bddb21130ada911445f20d70c3e2433b/lib/assets/javascripts/cartodb3/data/layer-definitions-collection.js#L213-L215
).

2. Right before the layerDefinitionsCollection is saved, the frontend checks if any of the layerDefinitionModels point to an analysis definition that doesn't exist yet. If that's the case, the above mentioned request is triggered. Only when all analysis definitions have been created successfully, the layerDefinitions are saved in bulk. Code is [here](https://github.com/CartoDB/cartodb/blob/94ee96f3bddb21130ada911445f20d70c3e2433b/lib/assets/javascripts/cartodb3/data/layer-definitions-collection.js#L62-L71).

@viddo raised some concerns about this, as some of this requests could fail and leave the app in an invalid state, or we'd need to add some logic to handle these cases. For that reason we talked about making this operations atomic.

Also, if a map has multiple layers, and new analysis definitions are need for all of them, we're now triggering one request for each of them (instead of saving them in bulk as we do with layers). 

**It's working fine right now so we can wait until ☝️ become actual problems**, but opening this issue to discuss about how we can go about this.

For instance, analysis definitions could be sent together with the layer information? Instead of doing:

```
{
  ...
  "options": {
    ...
    "source": "b0"
    ...
  },
  ...
}
```

The frontend could send:

```
{
  ...
  "options": {
    ...
    "source": {
      "id": "b0",
      "type": "source",
      "params": {
        "query": "SELECT * FROM untitled_table_116"
      },
      "options": {
        "table_name": "untitled_table_116"
      }
    },
    ...
  },
  ...
}
```

It'd then be the backend the one that determines if the analysis definition already exists or needs to be created, every time a layer is saved...

@viddo @juanignaciosl @xavijam @javierarce @javisantana thoughts?

**Comment by juanignaciosl** [comment]
I recall discussing about keeping analyses outside layers for a reason, but my memory is really bad.

I _think_ that we kept them separated because, from a conceptual point of view, an analyses doesn't require a layer to exist, and moving them inside would couple them tightly.

Deleting analyses might be an issue. When should backend delete an analysis? Sending an explicit DELETE doesn't fit the analyses-inside-layers very well, but not doing it doesn't provide enough information unless backend can know for sure that all analyses of a layer are always sent. Currently we're not even storing the layer id of an analysis, and we should.

These are my first, quick thoughts. I would really love rewinding to the conversations we had about this a couple of months ago, maybe somebody else can. I can't tell whether this is a clear win or not. It feels to me like changing some problems with others, but I might be wrong. You (frontenders) are the ones who are actually struggling with this, so it's up to you.

**Comment by viddo** [comment]
Thx for taking the time to write the issue, @alonsogarciapablo 

> @juanignaciosl I __think that we kept them separated because, from a conceptual point of view, an analyses doesn't require a layer to exist, and moving them inside would couple them tightly.

The first part is correct, and arguably the second one, too.

That said, while we could argue back and forth regarding the concepts and how they should be modeled, the main issue (at least from my point-of-view) is the lack of control of state mutations/lack of error-handling. When a state change involves more than one object, each save/state change could lead to erroneous state, and "rolling back" by doing more requests (which themselves could fail…) is far from ideal.  It's becoming even more obvious with #7878 (drag'n'drop nodes to create a new layer), since that involves both creating new layer/analysis and updating existing layer(s).

It would be good to have a conversation and decide how we want to solve this.

**Comment by viddo** [comment]
As of #8042 this afternoon I managed to pull out the implicit side-effects of listeners to a common place, so it's slightly better controlled), but proper error-handling/recovery is still missing. I'd like to have a conversation about this promptly to address this issue.
